packages/beacon-node/src/chain/historicalState/getHistoricalState.ts

```diff
@@ -0,0 +1,56 @@
+import {
+  BeaconStateTransitionMetrics,
+  CachedBeaconStateAllForks,
+  DataAvailableStatus,
+  ExecutionPayloadStatus,
+  PubkeyIndexMap,
+  createCachedBeaconState,
+  stateTransition,
+} from "@lodestar/state-transition";
+import {SignedBeaconBlock} from "@lodestar/types/allForks";
+import {BeaconConfig} from "@lodestar/config";
+import {IBeaconDb} from "../../db/index.js";
+
+export async function getClosestState(
+  slot: number,
+  config: BeaconConfig,
+  db: IBeaconDb
+): Promise<CachedBeaconStateAllForks> {
+  const states = await db.stateArchive.values({limit: 1, lte: slot, reverse: true});
+  if (!states.length) {
+    throw new Error("No close state found in the database");
+  }
+  return createCachedBeaconState(states[0], {
```

Comment:
creating a cached beacon state takes > 30s due to pubkey cache, the total memory for it is > 700MB. Is it possible to run state transition without using CachedBeaconState @wemeetagain ? I mean using `BeaconStateAllForks`

Reply:
I don't think we have the naive state transition code anymore. And I don't think we'd want to restore that for this usecase.

Maybe we can initialize the worker with a pubkey cache that we keep updated.

Reply:
We don't need the pubkey cache for historical states since we should not verify signatures. We can refactor the pubkey cache into being lazy:
- for normal operation pre-populate it explicitly
- for historical states worker do not prepopulate

Reply:
Luckily we already have this feature. We will still need the `pubkey2index`, but not the more costly `index2pubkey`.